### PR TITLE
Feature/luc/learn factory

### PR DESCRIPTION
**Pull request by lucpeterson**
This PR changes some of the sklearn-based scripts.
The key add is the creation of a factory for accessing any sklearn regressor available

```
>>> from ml import surrogates
>>> surrogates.sklearnRegressors.factory('GaussianProcessRegressor')
GaussianProcessRegressor(alpha=1e-10, copy_X_train=True, kernel=None,
             n_restarts_optimizer=0, normalize_y=False,
             optimizer='fmin_l_bfgs_b', random_state=None)
>>> surrogates.sklearnRegressors.factory('RandomForestRegressor',n_estimators=5)
RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=5, n_jobs=None,
           oob_score=False, random_state=None, verbose=0, warm_start=False)
```

I also changed learn.py to take advantage of this, so you can now do 
```
python learn.py -infile results_test.npz -regressor GaussianProcessRegressor -outfile GP42.pkl -reg_args '{"alpha":1e-9,"random_state":42}'
```

### CODE REVIEW

File: ml/surrogates.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        return _all_regressors
+
+    all_regs = reg_dict()
+
+    @classmethod
+    def factory(cls, name, *args, **kwargs):
+        if name in cls.all_regs:
+            return cls.all_regs[name](*args, **kwargs)
+        else:
+            print("Unknown regressor name "+name+"! For valid choices see sklearnRegressors.names()")
```
Comment (jwhite242) [inline ml/surrogates.py:34]: I'd recommend raising an exception here instead of just printing, that way learn.py can fail with more helpful error messages instead of trying to do fitting with the wrong object when this else gets hit.

File: ml/learn.py [hunk 2]
```diff
@@ -96,18 +92,20 @@ def stack_arrays(data, delimited_names, delimiter=','):
 
 def setup_argparse():
     parser = argparse.ArgumentParser(
-        description='Use sklearn to make a random forest regressor')
+        description='Use sklearn to make a regressor')
     parser.add_argument('-infile', help='.npz file with X and y data', default='results.npz')
```
Comment (ben-bay) [inline ml/learn.py:96]: How are you currently generating this `.npz` file? Don't have one to test with at the moment

Comment (ben-bay) [reply ml/learn.py:96]: Maybe we can write a function that generates an example `.npz` file to test on

File: ml/learn.py [hunk 3]
```diff
@@ -34,18 +34,11 @@
     import cPickle as pickle
```
Comment (ben-bay) [inline ml/learn.py:34]: Since merlin doesn't support python2, should we drop python2 support for merlin-spellbook as well? If so, I think this is a py2 pattern

File: ml/learn.py (hunk 2 above)
Comment (ben-bay) [reply ml/learn.py:96]: Though this may fall outside the scope of learn.py

File: ml/learn.py (hunk 3 above)
Comment (jwhite242) [reply ml/learn.py:34]: does merlin support running python2 in the steps still (in place of bash), or is that python3 only now too?

File: ml/learn.py (hunk 2 above)
Comment (jwhite242) [reply ml/learn.py:96]: probably be useful to have a unit test with for it, so not sure it's out of scope

File: ml/learn.py (hunk 3 above)
Comment (ben-bay) [reply ml/learn.py:34]: Good question, python2 shell does work, so I guess we should support it

Comment (jwhite242) [reply ml/learn.py:34]: so, going the other way, do we want to support that either?  most of the 3rd party libraries have dropped python2.  looks like scikit-learn 0.20 is the last one that works in python 2, and that was released in 2018.

File: ml/surrogates.py (hunk 1 above)
Comment (lucpeterson) [reply ml/surrogates.py:34]: good idea: will do

File: ml/learn.py (hunk 2 above)
Comment (lucpeterson) [reply ml/learn.py:96]: in practice, i'm doing it with the translator.py in that same folder. adding tests to these components (as well as stringing them together) is a good idea. i have to brush up on writing unit tests, but need to get this in production pretty fast. i'll either do it or accrue the technical debt

Comment (ben-bay) [reply ml/learn.py:96]: Happy to pay the debt once it's ready, can use tox/pytest the same as merlin does.

File: ml/learn.py (hunk 3 above)
Comment (ben-bay) [reply ml/learn.py:34]: My understanding is the merlin `shell` key is pretty open-ended. It will run a step in any shell you point at, so py2 isn't actually hardcoded in or explicitly supported

Comment (koning) [reply ml/learn.py:34]: This can be changed to:
try:
  import cPickle as pickle
except ImportError:
  import pickle

But the more fundamental question is will this module work if a user forces the py2 downgrade to a py2 sklearn module?

Comment (lucpeterson) [reply ml/learn.py:34]: the try ... except is already there to catch py2 issues

the functionality that this relies on is the sklearn "all_estimators" module, which has moved around some.
the try .. catch logic in surrogates.py catches older versions at least back to 0.15

File: ml/surrogates.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+from __future__ import print_function
+
+try:
+    from sklearn.utils import all_estimators
```
Comment (lucpeterson) [inline ml/surrogates.py:35]: this logic catches older versions of sklearn

Comment (lucpeterson) [reply ml/surrogates.py:35]: it looks like this library goes back at least 8 years
